Form field label (cc0f0a) update rule (#973)
act-rules/act-rules.github.io: _rules/form-control-label-descriptive-cc0f0a.md

  - DOM Tree
  - CSS Styling
authors:
  - Jean-Yves Moyen

> previous_authors:

Reviewer: Did we decide to do this? Have we documented this?

Currently the site will not enumerate/display this.

Author: There has been an informal decision that previous authors should be credited when applicable, without a formal decision on how to implement it precisely.

I agree that this is currently not displayed.

I wanted to keep the information in an hopefully not to bad way until this is correctly tackled.

Issue for this is already opened: #799 (from the earlier version of this PR…) now moved to https://github.com/act-rules/act-rules-web/issues/35 in the website repo.

  - Dagfinn Rømen
  - Geir Sindre Fossøy
---

## Applicability

This rule applies to any HTML or SVG element that:

- has one of the following [semantic roles][semantic role]: `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`; and
- is [visible][]; and
- is [included in the accessibility tree][]; and

> - has a [label][].

Reviewer 2: I do think the definition is fairly unambiguous as I'm at least having a hard time coming up with more than one way to interpret it. It is however far from objective, which is indeed an issue as we then cannot rely on it for the applicability of a rule.

> - has a [label][].

Author: As per #986, an assumption has been added that "WCAG label" should be visible and that the lack of equivalent accessible name is failing 4.1.2 instead. Thus it should be possible to focus on `label` elements and their context.

**Note**: The list of applicable [semantic roles][semantic role] is derived by taking all the roles from [WAI-ARIA Specifications](#wai-aria-specifications) that:

- inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and
- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.

**Note:** The `option` role is not part of the list of applicable roles, because it has a required context role that inherits from the `select` role. Furthermore, `option` does not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). This means that [WCAG 2.1][wcag21] does not require it to have an [accessible name][].

**Note**: [Labels][label] are not restricted to the `label` element of HTML and can be any element.

**Note**: [Labels][label] have to be presented to all users. Therefore, [accessible names][accessible name] or elements that are not [included in the accessibility tree][] do not qualify as [labels][label] (unless coupled with similar content provided for other users).

## Expectation

> The [label][] describes the purpose of the test target. The [context][] of the [label][] can be used to differentiate the purpose from other test targets in the same test subject.

Reviewer 3: > of the test target

Should be:

> of each test target

> The [label][] describes the purpose of the test target. The [context][] of the [label][] can be used to differentiate the purpose from other test targets in the same test subject.

Reviewer 3: This "the context of the label can be used" phrasing is ambiguous. What does that mean? "The label describes the purpose" doesn't say anything about the purpose needing to be differentiated. It needs to be described. There is no requirement that two things can't have the same label.

I would suggest:

> The label describes the purpose of the test target within its context.

**Note**: If the [label][] is not associated in a [programmatically determinable][] way with the form field element, this is likely a violation of [Success Criterion 1.3.1: Info and Relationships][sc131] but neither of this rule nor of [Success Criterion 2.4.6: Heading and Labels][sc246].

**Note**: It is possible for a test target to have an [accessible name][] without having an accurate [label][]. In that case, it would pass [Success Criterion 4.1.2: Name, Role and Value][sc412] but still fail this rule and [Success Criterion 2.4.6: Heading and Labels][sc246].

**Note**: While the [label][] has to be presented to all users, this is not necessarily the case for the [context][], if any. However, if [context][] is needed by some users to differentiate the purpose from one test target to another, it is likely that similar [context][] will be needed for all users.

> **Note**: Having a [label][] which is not included in the [accessible name][] is a violation of [Success Criterion 2.5.3: Label in Name][sc253] but neither of this rule nor of [Success Criterion 2.4.6: Heading and Labels][sc246].

Reviewer 3: > but neither of this rule nor of 

Should be:

> but not for this rule nor for

  - CSS Styling
acknowledgements:
  authors:
    - Jean-Yves Moyen
    - Wilco Fiers
  previous_authors:
    - Dagfinn Rømen
    - Geir Sindre Fossøy
---

## Applicability

> This rule applies to any HTML `label` element or other element referenced by `aria-labelledby` that is [visible][], and is programmatically associated with an HTML element which is [visible][] and has one of the following [semantic roles][semantic role]:

Reviewer 3: I think we can simplify this by quite a bit (and solve #1022 in the process) by creating a new definition:

```suggestion
This rule applies to any [visible][] HTML element that is a [programatic label][] for another HTML element that has one of the following [semantic roles][semantic role]:
```

And then add this as a definition (will need some word smithing and linking out to specs):

```md
---
name: programatic label
key: programatic-label
---

Element A is a _programatic label_ of element B, if both elements are [visible][], and one of the following is true:

1. element A is a label element that is implicitly or explicitly associated with element B; or
2. Element A is referenced by ID in the `aria-labelledby` attribute of element A
```

- `textbox`

and where both the element and the [programmatic label][] are [visible][].

**Note**: The list of applicable [semantic roles][semantic role] is derived by taking all the roles from [WAI-ARIA Specifications][] that:

- inherit from the [abstract][] `input` or `select` role; and
- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.

**Note:** The `option` role is not part of the list of applicable roles, because it has a required context role that inherits from the `select` role. Furthermore, `option` does not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). This means that [WCAG 2.1][wcag21] does not require it to have an [accessible name][].

**Note**: [Labels][label] in WCAG are not restricted to the `label` element of HTML and can be any element. This rule is only concerned about actual `label` elements, and elements that are programmatically marked as [labels][label] via the `aria-labelledby` attribute.

## Expectation

Each test target, together with its [visual context][], describes the purpose of the associated element.

> **Note**: It is possible for an element to have an [accessible name][] but still having a non-descriptive `label` element (and even a non-descriptive [label][]). In that case, it would pass [Success Criterion 4.1.2: Name, Role and Value][sc412] but still fail this rule and [Success Criterion 2.4.6: Heading and Labels][sc246].

Reviewer 4: ```suggestion
**Note**: It is possible for an element to have an [accessible name][] but still have a non-descriptive `label` element (and even a non-descriptive [label][]). In that case, it would pass [Success Criterion 4.1.2: Name, Role and Value][sc412] but still fail this rule and [Success Criterion 2.4.6: Heading and Labels][sc246].
```

**Note**: [Labels][label] in WCAG are not restricted to the `label` element of HTML and can be any element. This rule is only concerned about actual `label` elements, and elements that are programmatically marked as [labels][label] via the `aria-labelledby` attribute.

## Expectation

Each test target, within its [visual context][], describes the purpose of the associated element.

**Note**: It is possible for an element to have an [accessible name][] but still having a non-descriptive `label` (and even a non-descriptive [label][]). In that case, it would pass [Success Criterion 4.1.2: Name, Role and Value][sc412] but still fail this rule and [Success Criterion 2.4.6: Heading and Labels][sc246].

**Note**: Having a [label][] which is not included in the [accessible name][] is a violation of [Success Criterion 2.5.3: Label in Name][sc253] but not of this rule nor of [Success Criterion 2.4.6: Heading and Labels][sc246].

## Assumptions

> This rule assumes that [labels][label] are intended for sighted users, and that hiding a [visible][] [label][] from assistive technologies, is a failure of [sc412][], but not of [sc246][].

Reviewer 3: Please put link text in the SCs.

**Note:** The `option` role is not part of the list of applicable roles, because it has a required context role that inherits from the `select` role. Furthermore, `option` does not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). This means that [WCAG 2.1][wcag21] does not require it to have an [accessible name][].

**Note**: [Labels][label] in WCAG are not restricted to the `label` element of HTML and can be any element. This rule is only concerned about actual `label` elements, and elements that are programmatically marked as [labels][label] via the `aria-labelledby` attribute.

## Expectation

> Each test target, within its [visual context][], describes the purpose of the associated element.

Reviewer 3: Instead of saying "within it's visual context", this should probably say "together with it's visual context".

> Each test target, within its [visual context][], describes the purpose of the associated element.

Reviewer 3: You need to define "associated element" here, or use something unambiguous. Keep in mind that a single label can be associated with multiple fields.

Author: Indeed, I completely missed the possibility for multiple association… 😞

#### Passed Example 1

The implicit `label` element acts as a [label][] and it is descriptive.

```html
<label>First name:<input id="fname" type="text" name="first_name"/></label>
```

#### Passed Example 2

The explicit `label` element acts as a [label][] and it is descriptive.

```html
<label for="fname">First name:</label> <input id="fname" type="text" name="first_name" />
```

#### Passed Example 3

The `span` element acts as a [label][] and it is descriptive. Even if it is not associated to the form field in a [programmatically determinable][] way, it is still presented to all users and thus qualifies as a [label][].

```html
<span>First name:</span> <input type="text" name="first_name" />
```

#### Passed Example 4

> Even through the `span` element is not [included in the accessibility tree][], it is associated to the form field in a [programmatically determinable][] way and gives it an [accessible name][]. Therefore, it is presented to all users and qualifies as a [label][].

Reviewer 5: through > though

```html
<h2>Shipping address</h2>
<label>Name<input id="shipping-name" type="text" name="name"/></label>
<label>Street<input id="shipping-street" type="text" name="street"/></label>

<h2>Billing address</h2>
<label>Name<input id="billing-name" type="text" name="name"/></label>
<label>Street<input id="billing-street" type="text" name="street"/></label>
```

### Failed

#### Failed Example 1

The implicit `label` element acts as a [label][] and it is not descriptive.

```html
<label>City:<input id="fname" type="text" name="first_name"/></label>
```

#### Failed Example 2

The explicit `label` element acts as a [label][] and it is not descriptive.

```html
<label for="fname">City:</label> <input id="fname" type="text" name="first_name" />
```

#### Failed Example 3

The `span` element acts as a [label][] and it is not descriptive. Even if it is not associated to the form field in a [programmatically determinable][] way, it is still presented to all users and thus qualifies as a [label][].

```html
<span>City:</span> <input type="text" name="first_name" />
```

#### Failed Example 4

> Even through the `span` element is not [included in the accessibility tree][], it is associated to the form field in a [programmatically determinable][] way and gives it an [accessible name][]. Therefore, it is presented to all users and qualifies as a [label][]. It is not descriptive.

Reviewer 5: through > though

<h2 style="position: absolute; top: -9999px; left: -9999px;">Billing address</h2>
<input aria-label="Name" id="billing-name" type="text" name="name" />
<input aria-label="Street" id="billing-street" type="text" name="street" />
```

### Inapplicable

#### Inapplicable Example 1

There is no element with any of the required [semantic roles][semantic role].

```html
<label for="fname">First name:</label>
<p id="fname"></p>
```

#### Inapplicable Example 2

The `label` element is [hidden](https://www.w3.org/TR/accname-1.1/#dfn-hidden) to all users and thus does not qualifies as a [label][].

```html
<label for="fname" style="display:none;">First name:</label> <input id="fname" type="text" name="first_name" />
```

#### Inapplicable Example 3

The form field is not [visible][].

```html
<label
	>First name: <input style="position: absolute; top: -9999px; left: -9999px;" type="text" name="first_name"
/></label>
```

#### Inapplicable Example 4

The form field is not [included in the accessibility tree][].

```html
<span>First name:</span><input aria-hidden="true" type="text" name="first_name" />
```

#### Inapplicable Example 5

> The `span` element is not [visible][]. Therefore, it is not presented to all users and does not qualify as a [label][]. The form field has no [label][], even though at has an [accessible name][].

Reviewer 5: at > it

> #### Passed Example 1

Reviewer 3: Most of these examples don't need to be changes. Changing them invalidates a lot of test data. Please keep the examples the way they are, except for changes related to the changes you made in applicability / expectations.
